tfjs-backend-webgpu/src/depthwise_conv2d_vec4_webgpu.ts

```diff
@@ -58,57 +60,66 @@ export class DepthwiseConv2DVec4Program implements WebGPUProgram {
     this.activation = activation;
     this.hasPreluActivation = hasPreluActivation;
 
-    this.shaderKey = `depthwiseVec4_${activation}_${
-        this.convInfo.filterHeight}_${this.convInfo.filterWidth}`;
+    this.shaderKey =
+        `depthwiseVec4_${activation}_${this.convInfo.filterHeight}_${
+            this.convInfo.filterWidth}_${this.convInfo.strideHeight}_${
+            this.convInfo.strideWidth}_${this.workPerThread}`;
   }
 
   getUserCode(): string {
-    // Here 4 is the work per thread in X dimension.
-    const xNumber = 4 + this.convInfo.filterWidth - 1;
+    const xNumber = (this.workPerThread - 1) * this.convInfo.strideWidth +
+        this.convInfo.filterWidth;
+
     const userCode = `
       ${activationFnSnippet(this.activation, this.hasPreluActivation, true, 4)}
       fn readX(batch : i32, row : i32, col : i32, channel : i32) -> vec4<f32> {
         var value = vec4<f32>(0.0);
-        if (row >=0 && row < uniforms.inDims[0] && col >=0 && col < uniforms.inDims[1])
+        if (col >=0 && col < uniforms.inDims[1])
         {
           value = getX(batch, row, col, channel);
         }
         return value;
       }
+
+      const strideHeight = ${this.convInfo.strideHeight};
+      const strideWidth = ${this.convInfo.strideWidth};
       ${getWorkGroupSizeString()}
       fn _start(@builtin(global_invocation_id) globalId: vec3<u32>) {
         let batch = i32(globalId.z) / uniforms.outShape[1];
         let r = i32(globalId.z) % uniforms.outShape[1];
-        let c = i32(globalId.y) * 4;
+        let c = i32(globalId.y) * ${this.workPerThread};
         let d1 = i32(globalId.x) * 4;
-        let xRCCorner = vec2<i32>(r, c) - uniforms.pad;
+        let xRCCorner = vec2<i32>(r, c) * vec2<i32>(strideHeight, strideWidth) - uniforms.pad;
 
         let xRCorner = xRCCorner.x;
         let xCCorner = xRCCorner.y;
         var xVals : array<vec4<f32>, ${xNumber}>;
-        var dotProd : array<vec4<f32>, 4>;
-        dotProd[0] = vec4<f32>(0.0);
-        dotProd[1] = vec4<f32>(0.0);
-        dotProd[2] = vec4<f32>(0.0);
-        dotProd[3] = vec4<f32>(0.0);
+        var dotProd : array<vec4<f32>, ${this.workPerThread}>;
+        for (var i = 0; i < ${this.workPerThread}; i++)
+        {
+          dotProd[i] = vec4<f32>(0.0);
+        }
 
         // Use constant instead of uniform can give better performance.
         for (var wR = 0; wR < ${this.convInfo.filterHeight}; wR = wR + 1) {
           let xR = xRCorner + wR;
-          for (var i = 0; i < ${xNumber}; i++)
+          if (xR >=0 && xR < uniforms.inDims[0])
```

Comment:
Since you moves `xR`'s boundary checks out of `readX` function, should we also move `col`'s boundary check out of `readX` function?

Reply:
We can move `xR`'s boundary checks out of `readX` is because we are processing the values in the same row. However, we need to check whether each value is out of boundary for `col`. So `col`'s boundary check is necessary for each value.

Reply:
Yes. I mean,  in `readX`, it may be confusing that we only do boundary check for `col` while do not for `row`. Since `readX` is only used once and it's very straight forward, this LGTM and I was kindly reminding.

Approved.